cpp/src/arrow/array/array_nested.h

```diff
@@ -189,6 +189,11 @@ class ARROW_EXPORT ListArray : public BaseListArray<ListType> {
   Result<std::shared_ptr<Array>> Flatten(
       MemoryPool* memory_pool = default_memory_pool()) const;
 
+  /// \brief Flatten all level recursively until reach a non-list type, and return a
+  /// non-list type Array.
```

Comment:
I think we should be explicit and list the types that we consider as logical lists here. It should include lists, list-views, fixed-size-lists, and map.

<img width="1615" alt="Screenshot 2024-04-09 at 15 06 26" src="https://github.com/apache/arrow/assets/207795/e89cb9fc-a707-4d73-b83d-d20efca71ec7">

Reply:
And you only have to declare it on `VarLengthListLikeArray<>` and `FixedSizeListArray` -- they can all simply call the same non-inline function that dispatches based on `type->id()`. That extra check won't be significant for a function that scans the entire array and allocates.